docs/reference/providers/custom/building-custom-provider.md

```diff
@@ -0,0 +1,263 @@
+---
+title: Building Custom Providers
+description: How to build a custom provider using Nitric
+---
+
+This guide is for those that are interested in building their own custom provider to provide support for a missing cloud, or implementing their own internal development platform, or just curious to know how it all works. Nitric's main goal is to keep a general interface for interacting with cloud resources, regardless of the provider. This abstraction enables portability, developer efficiency, and standardizes the way that code is written across teams.
+
+Nitric's cloud providers are individual plugins separated into a runtime and a deployment provider for AWS, GCP, and Azure. The [runtime provider](#runtime) is concerned with converting abstract SDK calls into direct cloud API calls. The [deployment provider](#deployments) takes a resource specification of topics, buckets, and functions and converts them into cloud resources. The providers, CLI, and SDK all work together to make the application deploy and run.
+
+![Nitric Architecture](../../../../assets/img/guides/custom-provider/custom-provider-architecture.png)
+
+## Deployments
+
+The Nitric CLI's core responsibility is to get the infrastructure requirements from a Nitric application's code. The requirements are then composed into a deployment specification which the CLI and deployment providers both expect. This deployment specification is written using [protocol buffers](https://protobuf.dev/) and transmitted to a deployment provider using [gRPC](https://grpc.io/).
+
+> The complete proto files can be found in our core [nitric repository](https://github.com/nitrictech/nitric/blob/main/contracts/proto/deploy/v1/deploy.proto).
+
+The CLI will also containerize and build your function code, passing the information about the built image to the deployment provider. Given this deployment specification, the provider can deploy whatever resources are needed. For example, given the function below:
+
+```ts
+import { api, bucket, faas } from '@nitric/sdk';
+
+const helloApi = api('main');
+
+const buck = bucket('images').for('writing');
+
+helloApi.get('/hello/:name', async (ctx: faas.HttpContext) => {
+  const { name } = ctx.req.params;
+
+  ctx.res.body = `Hello ${name}`;
+
+  return ctx;
+});
+```
+
+The CLI would convert it into the following resource spec:
+
+```json
+{
+  "spec": {
+    "resources": [
+      {
+        "name": "images",
+        "type": 2,
+        "Config": {
+          "Bucket": {}
+        }
+      },
+      {
+        "name": "main",
+        "Config": {
```

Comment:
Why doesn't this API resouce have a `type` property like the other resources?

Reply:
It was getting removed in the JSON serialization because its `"type": 0`